### PR TITLE
x.py: Support systems with only `python3` not `python`

### LINKED ISSUE #71818: Rust doesn't build without "python" name being available
Since Ubuntu 20.04, the name `python` [no longer available by default](https://wiki.ubuntu.com/FocalFossa/ReleaseNotes#Python3_by_default) without `python-is-python2` package being installed. The build scripts should point to `python2` instead.

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #71818 without the pitfalls so far described in previous attempts.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I've never done much development on Windows, but is bash typically available there?

FWIW, I believe the current trajectory is to add an x.ps1 PowerShell script (not sure on extension, I think that's right?) which would be ~equivalent to the x.py wrapper, so even if python isn't available adding that might let us move forward regardless.

**Comment by est31** [comment]
cc also https://github.com/python/cpython/issues/90803

**Comment by ChrisDenton** [comment]
> I've never done much development on Windows, but is bash typically available there?

[msys2](https://www.msys2.org/) and [git bash](https://www.atlassian.com/git/tutorials/git-bash) are development platforms that emulates a nix environment on Windows, including common gnu tools and bash, In those environments this file will be run in bash.

"Native" Windows shells will simply treat this as a python file. So it will run whatever python version is associated with `.py` files. No bash involved.

**Comment by joshtriplett** [comment]
Does this need to use bash rather than sh?

File: x.py [hunk 1]
```diff
@@ -1,17 +1,38 @@
...
 # interpreter.
+#
+# `./x.py` would not normally benefit from this because the bash above tries
+# python3 before 2, but this matters if someone ran `python x.py` and their
+# system's `python` is python2.
 if sys.version_info.major < 3:
     try:
-        # On Windows, `py -3` sometimes works.
-        # Try this first, because 'python3' sometimes tries to launch the app
-        # store on Windows
```
Comment (jyn514) [inline x.py:14]: Can you please add back this comment?

**Comment by jyn514** [comment]
I tried this in https://github.com/rust-lang/rust/pull/80585 and it broke because powershell didn't know how to run it. Can you confirm this works on windows with powershell?

I had other issues there, but I think they might have been because I intentionally omitted the shebang. IIRC the reason for omitting the shebang is because there's something called `py` on Windows that special cases the `/usr/bin/python` and `/usr/bin/python3` shebangs, and will give an error for anything else. Can you confirm this works when using `py x.py`, or py is the default interpreter for python files?

**Review by jyn514 — CHANGES_REQUESTED** [review]

**Comment by ChrisDenton** [comment]
> because there's something called `py `

Oh hm. Right the `py` launcher again.

If python is installed via the Microsoft store, winget, manually, etc then you don't get this "helpful" utility and everything works fine because Windows shells don't read shebangs.

If python is installed via the official installer from python.org then that will install the `py` launcher which does read shebangs (and therefore fails). The user could manually fix this by changing the file association (either to the real python or to run `py -3`). But if user invention is required anyway then that defeats the point of finding python automatically.

**Comment by jyn514** [comment]
@ChrisDenton to be clear, the issue isn't 2 vs 3 - x.py already handles trying to relaunch itself with 3 if necessary. The problem I hit was that `py` *only* understands the /usr/bin/python shebang, because it doesn't run in an msys environment and thinks /usr doesn't exist if you try `/usr/bin/env bash`.

**Comment by ChrisDenton** [comment]
@jyn514 Yes. Sorry I wasn't clear there. What I'm saying is that the installer sets `.py` files to be run by the `py.exe` launcher. This launcher, as you say, reads the file shebang. However, if the user changes the `.py` association to be run by `py.exe -3` then `py.exe` skips reading the shebang. They could also use `py.exe -2` (if they also have python 2 installed). This will also skip reading the shebang.

Or they could set the association to run `python.exe` directly and skip the middleman.

**Comment by dtolnay** [comment]
> Can you confirm this works on windows with powershell?

Not a proficient Windows user, but toying around in a VM, I was not able to get `./x.py` to run in powershell neither *before* nor *after* this change, neither with Python from Microsoft Store nor python.org. Here is what I found:

| setup | `./x.py`<br>works<br>before? | `./x.py`<br>works<br>after? | workaround |
|---|---|---|---|
| Linux distro pre-2020 | yes | yes | |
| Ubuntu 20.04+, other recent Linux | **no** | yes | install `python-is-python3` or similar,<br>or `python3 x.py` |
| macOS pre-2022 | yes | yes | |
| macOS Monterey+ | **no** | yes | `python3 x.py` (is there a thing to install?) |
| Windows, MSYS2 or git bash, any Python | yes | yes | |
| Windows Subsystem for Linux | yes | yes | |
| Windows, Powershell, Python from MS | **no** | **no** | set file association to python.exe,<br>or `python3 x.py` (`py -3` does not work) |
| Windows, Powershell, python.org | **no** | **no** | set file association to `py.exe -3`,<br>or `py -3 x.py` (`python3` does not work,<br>it opens Microsoft Store) |

File: x.py (hunk 1 above)
Comment (dtolnay) [reply x.py:14]: This comment got moved to line 8. I think it is sufficient to have it once.

**Comment by dtolnay** [comment]
I think my powershell result agrees with https://github.com/rust-lang/rust/issues/71818#issuecomment-1069175579: the shebang currently on master **already** doesn't work:

Screenshot 2022-06-25 at 10-05-46 Rust doesn't build without python name being available · Issue #71818 · rust-lang_rust

so before this PR we manage to be broken on *all* modern environments across Linux, macOS, and Windows. :sweat_smile:

**Comment by jyn514** [comment]
@wesleywiser @rylev @ChrisDenton you use Windows and PowerShell regularly, right? Do you normally use `./x.py`? Or `python3 x.py`?

**Comment by dtolnay** [comment]
I think @ChrisDenton answered that in https://github.com/rust-lang/rust/pull/98474#issuecomment-1166314485, and I found the same: `./x.py` works in Powershell (before and after PR identically) as long as you do the one-time setup of associating to run it with `py.exe -3`.

> *However, if the user changes the `.py` association to be run by `py.exe -3` then `py.exe` skips reading the shebang.*

**Comment by ChrisDenton** [comment]
The `x.exe` usually. But `./x` or `./x.py` also works for me.

**Review by jyn514 — APPROVED** [review]

**Comment by ChrisDenton** [comment]
Btw, here's a quick powershell log after installing from python.org in a fresh sandbox:

```
PS C:\Users\WDAGUtilityAccount> $env:PATHEXT+=';.PY'
PS C:\Users\WDAGUtilityAccount> cat x.py
#!/usr/bin/env bash

'''':
echo "hello from bash"
exit 0
'''

print("hello from powershell")
PS C:\Users\WDAGUtilityAccount> ./x
Unable to create process using '/usr/bin/env bash "C:\Users\WDAGUtilityAccount\x.py" '
PS C:\Users\WDAGUtilityAccount> py x.py
Unable to create process using '/usr/bin/env bash x.py'
PS C:\Users\WDAGUtilityAccount> py -3 x.py
hello from powershell
```

On the machine I use for development (using the same x.py as above):

```
PS C:\Users\Chris> ./x
hello from powershell
PS C:\Users\Chris> ./x.py
hello from powershell
PS C:\Users\Chris> python x.py
hello from powershell
```

It's possible I did some additional setup though. I'm using the python from the Windows store.

**Comment by Mark-Simulacrum** [comment]
> I think @ChrisDenton answered that in [#98474 (comment)](https://github.com/rust-lang/rust/pull/98474#issuecomment-1166314485), and I found the same: `./x.py` works in Powershell (before and after PR identically) as long as you do the one-time setup of associating to run it with `py.exe -3`.
> 
> > _However, if the user changes the `.py` association to be run by `py.exe -3` then `py.exe` skips reading the shebang._
> 
> In the "Administrator Command Prompt": `ftype Python.File="C:\Windows\py.exe" -3 "%L" %*`

I think it'd be great to get this line noted somewhere in our documentation -- maybe the rustc-dev-guide at least? We could consider emitting a one-line command if it's not the case (presumably detectable?) from x.py or similar.

In the meantime I'm happy with the table (https://github.com/rust-lang/rust/pull/98474#issuecomment-1166324726) and other comments here sufficiently to let us move forward; we can revisit if this causes regressions in practice. Thanks @dtolnay for pushing this forward!

@bors r+ rollup

**Comment by dtolnay** [comment]
I can follow up with a docs PR. I know almost nothing about Windows beyond what's discussed above and in the issue, so maybe it would be even better if someone regularly doing development on Windows could handle the docs.

Here is a draft:

---

**PowerShell users:** if `./x.py` gives you an error saying _"Unable to create process using '/usr/bin/env bash …'"_ then:

- If `py -3 x.py` seems to run successfully, you can run `ftype Python.File="C:\Windows\py.exe" -3 "%L" %*` in the "Administrator Command Prompt" to make `x.py` by itself work.

- If `py -3 x.py` does not work but `python3 x.py` does, you can run <kbd><kbd>????</kbd></kbd> to make `x.py` by itself work.

---

Maybe we could put this not just in the rustc-dev-guide but also directly into x.py, since that's probably where I would first look after a failure.

**Comment by wesleywiser** [comment]
> @wesleywiser @rylev @ChrisDenton you use Windows and PowerShell regularly, right? Do you normally use `./x.py`? Or `python3 x.py`?

@jyn514 I use `python x.py` from Powershell.

**Comment by CAD97** [comment]
This broke `x.py` on Windows machines. I've been using `py ./x.py` without trouble until my most recent `git pull`.

**Comment by CAD97** [comment]
In general, it seems suspicious to have a `.py` with a bash shebang. imho this should've been a `x.sh`.

**Comment by Mark-Simulacrum** [comment]
I'd prefer to avoid a revert without more information - at the very least, to determine what we can do to test in the future, given that comments on this thread suggest to me that the file should continue to work with py x.py...

**Comment by jyn514** [comment]
@Mark-Simulacrum I think @dtolnay tested with ./x.py, but not `py x.py`.

**Comment by Mark-Simulacrum** [comment]
Ah, I see. https://github.com/rust-lang/rust/pull/98474#issuecomment-1166332721 has the bash error that's presumably not working - py x.py doesn't work, but apparently py -3 x.py does, and presumably `ftype Python.File="C:\Windows\py.exe" -3 "%L" %*` then works.

**Comment by CAD97** [comment]
I can confirm that the status on my install is (from cmd)

```
D:\git\rust-lang\rust>winget list python
Name            Id                                     Version     Available   Source
-------------------------------------------------------------------------------------
Python 3        Python.Python.3                        3.10.4150.0 3.10.5150.0 winget
Python Launcher {691AAAA1-FE86-4973-8DA2-6AA2B3327562} 3.10.7751.0

D:\git\rust-lang\rust>ftype | rg py
Python.ArchiveFile="C:\Windows\py.exe" "%L" %*
Python.CompiledFile="C:\Windows\py.exe" "%L" %*
Python.File="C:\Windows\py.exe" "%L" %*
Python.NoConArchiveFile="C:\Windows\pyw.exe" "%L" %*
Python.NoConFile="C:\Windows\pyw.exe" "%L" %*

D:\git\rust-lang\rust>py ./x.py --help
Unable to create process using '/usr/bin/env bash ./x.py --help'

D:\git\rust-lang\rust>x.py --help
Unable to create process using '/usr/bin/env bash "D:\git\rust-lang\rust\x.py"  --help'

D:\git\rust-lang\rust>py -3 ./x.py --help
info: Downloading and building bootstrap before processing --help
      command. See src/bootstrap/README.md for help with common
      commands.
Building rustbuild
    Finished dev [unoptimized] target(s) in 0.06s
Usage: x.py <subcommand> [options] [<paths>...]
```

I did install python through winget, but I installed [`Python.Python.3`](https://winget.run/pkg/Python/Python.3), not the msstore python. I also manually uninstalled the msstore `python3` shim.

Note that on Windows, the story about when you have to use `./x.py` or `x.py` is... inconsistent. cmd requires `x.py`; `./x.py` will fail to find a program `.`. pwsh requires `./x.py`, and `x.py` will not work unless you add `.` to `$env:PATH`, to avoid pwd files overriding programs loaded from `PATH`.

Because of this, even if I change my `.py` file handler, my muscle memory is still to call `py ./x.py`, because this works both in pwsh and cmd (and nushell, what I'm actually daily driving atm, uses cmd to run externals, so behaves like cmd).

**Comment by CAD97** [comment]
I opened an issue: #98650

**Comment by jyn514** [comment]
I think we should revert this change.

# Before this PR

## Unix

You need `python` in PATH. The errors when using ./x.py on Linux are fairly reasonable ("command python not found"). People sometimes run into issues when they first start contributing, but they're easy to work around.

## Windows

- ./x.py doesn't work because there's no program associated with it
- py, python3, python2 all work.

# After this PR

## Unix

You need bash in path, and any of python3, python, or python2 installed.

## Windows

- py breaks with unhelpful errors ("/usr/bin/env not found"). This breaks existing workflows, not new contributors.
- ./x.py still doesn't work.

**Comment by ChrisDenton** [comment]
For completeness I'd add that installing the `py.exe` launcher is optional (though default) in the python.org installer.

**Pull request merged**